pkg/fleetautoscalers/fleetautoscalers.go

```diff
@@ -224,3 +234,425 @@ func applyBufferPolicy(b *autoscalingv1.BufferPolicy, f *agonesv1.Fleet) (int32,
+		return 0, false, err
+	}
+
+	for _, gs := range gameServers {
+		replicas--
+		switch isCounter {
+		case true:
+			if counter, ok := gs.Status.Counters[key]; ok {
+				aggCount -= counter.Count
+				aggCapacity -= counter.Capacity
+			}
+		case false:
+			if list, ok := gs.Status.Lists[key]; ok {
+				aggCount -= int64(len(list.Values))
+				aggCapacity -= list.Capacity
+			}
+		}
+		availableCapacity = aggCapacity - aggCount
+		// Check if we've hit MinCapacity
+		if aggCapacity == minCapacity {
+			return replicas, true, nil
+		}
+		if aggCapacity < minCapacity {
+			return replicas + 1, true, nil
+		}
+		// Check if we're at our desired Buffer
+		if availableCapacity == buffer {
+			return replicas, false, nil
+		}
+		if availableCapacity < buffer {
+			return replicas + 1, false, nil
+		}
+	}
+
+	if replicas < 0 { // This shouldn't ever happen, but putting it here just in case.
+		replicas = 0
+	}
+
+	return replicas, false, nil
+}
+
+// scaleUpByPercent
+func scaleUpByPercent(currReplicas int32, count, aggCount, capacity, aggCapacity,
+	maxCapacity int64, desiredCapacity, bufferPercent float64) (int32, bool, error) {
+
+	if capacity == 0 {
+		return 0, false, errors.Errorf("cannot scale up as Capacity is equal to 0")
+	}
+
+	additionalReplicas := int64(math.Ceil((desiredCapacity - float64(aggCapacity)) / (float64(capacity) - float64(count))))
+	replicas := currReplicas + int32(additionalReplicas)
+	// Case where adding a List or Counter does not change the Count (and thus does not change our desiredCapacity)
+	if count == 0 {
+		return replicas, false, nil
+	}
+
+	// In case where len(list.Values) or counter.Count != 0 then we need to update desiredCapacity
+	// each time we add a List or Counter. Start at point where desired replicas was determined based
+	// on case where len(list.Values) or counter.Count == 0.
+	aggCount += count * additionalReplicas
+	aggCapacity += capacity * additionalReplicas
+	limited := false
+	for {
+		// Check if we've reached MaxCapacity
+		if aggCapacity == maxCapacity {
+			limited = true
+			break
+		}
+		if aggCapacity > maxCapacity {
+			limited = true
+			replicas--
+			break
+		}
+		// Check if we've reached desiredCapacity
+		desiredCapacity = (float64(aggCount) * 100) / (100 - bufferPercent)
+		desiredReplicas := math.Ceil(desiredCapacity / float64(capacity))
+		if replicas >= int32(desiredReplicas) {
+			break
+		}
+		// Keep checking if adding one List or Counter will reach our desiredCapacity
+		aggCount += count
+		aggCapacity += capacity
+		replicas++
+	}
+
+	return replicas, limited, nil
+}
+
+// scaleDownByPercent
+// TODO: This death-spirals down to min capacity, so the customer would need to have something
+// in place to prevent gameservers from being evicted elsewhere if they have Count / Values on
+// them. I'm assuming this is the behavior we intend, and it just needs good documentation?
+func scaleDownByPercent(f *agonesv1.Fleet, gameServerLister listeragonesv1.GameServerLister,
```

Comment:
// TODO: This death-spirals down to min capacity, so the customer would need to have something
// in place to prevent gameservers from being evicted elsewhere if they have Count / Values on
// them. I'm assuming this is the behavior we intend, and it just needs good documentation?

Reply:
This is what we worked out on our doc. Allocated GameServers never go away, so their list/counts are counted towards the amount that is needed - so it shouldn't death spiral to min, it should do as expected (least I hope so - that exercise made my head hurt 🤯 )

Reply:
It will death spiral to minCapacity -- all of the bufferPolicy, etc. calculate the "desired replicas" and it's not until the computeDesiredFleetSize returns back to the controller that the controller determines how many replicas can actually be deleted.

Reply:
Resolved with refactor.